### ISSUE TITLE
Threshold to be set 0 when using binary accuracy with raw prediction values (from_logits=True)

### ISSUE THREAD

**Issue by HaraBeron** [tensorflow/tensorflow#41413]
This is mainly a documentation bug (official tensorflow tutorial), but it is a "dangerous trap" and might also happen in general to users, so see below my last sentence this could also be fixed in Tensorflow that it detects this automatically.

In this [tutorial](https://www.tensorflow.org/tutorials/images/transfer_learning) raw prediction values (form_logit=True) are used. So we have negative values and positive values, while 

> 
"prediction will be treated as a logit, or a raw prediction value. **Positive numbers predict class 1, negative numbers predict class 0.**"

However, the model.compile statement is as follows:

```
base_learning_rate = 0.0001
model.compile(optimizer=tf.keras.optimizers.RMSprop(lr=base_learning_rate),
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

```

This is wrong, as per default, threshold value to classify is 0.5:

https://www.tensorflow.org/api_docs/python/tf/keras/metrics/binary_accuracy
https://www.tensorflow.org/api_docs/python/tf/keras/metrics/BinaryAccuracy
```
tf.keras.metrics.binary_accuracy(
    y_true, y_pred, threshold=0.5
)
```
```
tf.keras.metrics.BinaryAccuracy(
    name='binary_accuracy', dtype=None, threshold=0.5
)
```

> threshold | (Optional) Float representing the threshold for deciding whether prediction values are 1 or 0.
> -- | --

This leads to the wrong classifications. model.evaluate will also give false accuracy measure. Reason is that predicted values in range [0,0.49999] are wrongly classified as 0 (I am not sure what happens to a value of exactly 0.5), whereas they actually should be classified as 1!

So it needs to be corrected to:

> model.compile(optimizer=tf.keras.optimizers.RMSprop(lr=base_learning_rate),
>               loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
>               metrics=tf.keras.metrics.BinaryAccuracy(threshold=0.0))

Would be even better if this is corrected inside Tensorflow that it automatically detects that from_logits=True was set and then assumes that default threshold is not 0.5 anymore, but 0.0 (and maybe additional WARNING output).

**Comment by MarkDaoust** [comment]
This is fixed.

**Comment by thariq-nugrohotomo** [comment]
Hi @MarkDaoust @HaraBeron ,

Seems like the fixes are lost after being replaced by this commit https://github.com/tensorflow/docs/commit/adae1941b3a1b3c0b781b94d96f4476892a44620

Could you please reopen this issue?

Also, rather than only fixing the documentation, I would consider to add a new `from_logits` argument for `BinaryAccuracy`, and set the default `threshold` to `None` (for automatically decide whether to use `0.5` or `0.0`)

**Comment by PrimozGodec** [comment]
Hi @MarkDaoust  and others, when I am looking at https://www.tensorflow.org/tutorials/images/transfer_learning#compile_the_model_2, there is still the `"accuracy"` metric used instead of `tf.keras.metrics.BinaryAccuracy(threshold=0)` when `from_logits=True`. Am I looking wrong, or is "accuracy" now setting the threshold according to `from_logits` parameters?

**Comment by MarkDaoust** [comment]
Keras is careful about building a correct metric object when you pass "accuracy".

The code is here:

https://github.com/tensorflow/tensorflow/blob/631d5829e51d7401df514ab6ea20128711708d80/tensorflow/python/keras/engine/compile_utils.py#L503-L528

The output in the notebook looks right. It couldn't be if Keras had built the wrong metric object. 

The initial report may have been over cautious.

I believe the code is correct as is.
